db/fixtures/ae_datastore/ManageIQ/Deployment/ContainerProvider/System/StateMachine/Deployment.class/__methods__/pre_deployment.rb

```diff
@@ -1,93 +1,53 @@
+    deployment.subscribe_deployment_master(deployment_master, user)
+  end
+  deployment.perform_agent_commands(deployment_master, user, commands)
+  if release.include?("Red Hat Enterprise Linux") && deployment.nodes_subscription_needed?
+    deployment.subscribe_cluster(deployment_master, user)
+    $evm.root['ae_result']         = 'retry'
+    $evm.root['ae_retry_interval'] = '1.minute'
+  else
+    $evm.root['ae_result']               = "ok"
+    $evm.root['automation_task'].message = "#{$evm.root['ae_state']} was finished successfully"
   end
-  ssh.loop
-  {
-    :stdout      => stdout_data,
-    :stderr      => stderr_data,
-    :exit_code   => exit_code,
-    :exit_signal => exit_signal
-  }
 end
 
-def pre_deployment
-  commands = ['sudo yum install -y ansible',
-              'sudo yum install -y openshift-ansible openshift-ansible-playbooks pyOpenSSL',
-              "sudo mv ~/inventory.yaml /usr/share/ansible/openshift-ansible/"
-              ]
-  $evm.log(:info, "********************** #{$evm.root['ae_state']} ***************************")
-  begin
-    Net::SSH.start($evm.root['deployment_master'], $evm.root['user'], :paranoid => false, :forward_agent => true,
-                   :key_data => $evm.root['private_key']) do |ssh|
-      $evm.log(:info, "Connected to deployment master, ip address: #{$evm.root['deployment_master']}")
-      system "scp -o 'StrictHostKeyChecking no' inventory.yaml #{$evm.root['user']}@#{$evm.root['deployment_master']}:~/"
-      failed_execute = false
-      release = ssh.exec!("cat /etc/redhat-release")
-      if release.include?("CentOS")
-        commands.unshift("sudo yum install epel-release -y",
-                         "sudo curl -o /etc/yum.repos.d/maxamillion-origin-next-epel-7.repo #{REPO_URL}")
-      elsif release.include?("Red Hat Enterprise Linux") &&
-            !$evm.root['automation_task'].automation_request.options[:attrs][:containerized]
-        commands = handle_rhel_subscriptions(commands)
-      end
-      commands.each do |cmd|
-        result = ssh_exec!(ssh, cmd)
-        unless result[:exit_code] == 0
-          $evm.root['automation_task'].message = "FAILED: couldn't execute command #{cmd}. ERROR: #{result[:stderr]}"
-          $evm.root['ae_result'] = "error"
-          failed_execute = true
-        end
-        break if failed_execute
-      end
-      unless failed_execute
-        $evm.root['ae_result'] = "ok"
-        $evm.root['automation_task'].message = "successful pre-deployment"
-      end
+begin
+  $evm.root['container_deployment'] = $evm.vmdb(:container_deployment).find(
+    $evm.root['automation_task'].automation_request.options[:attrs][:deployment_id]) unless $evm.root['container_deployment']
+  user = $evm.get_state_var(:ssh_user)
+  deployment_master = $evm.root['deployment_master']
+  deployment = $evm.root['container_deployment']
+  if deployment.playbook_running?
+    result = deployment.subscribe_cluster(deployment_master, user)
```

Comment:
Can you explain the flow here?
When was a playbook launched?  Why if yes are you subscribing "again"?
(Does `subscribe_cluster` just poll status if it's already running?  Isn't there a race condition where if just finished it would start again?)